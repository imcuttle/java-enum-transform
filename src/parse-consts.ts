import {
  CstNode,
  BaseJavaCstVisitorWithDefaults,
  parse,
  ClassDeclarationCtx,
  EnumDeclarationCtx,
  EnumConstantCtx,
  ExpressionCtx,
  IntegerLiteralCstNode,
  IntegerLiteralCtx,
  IToken
} from 'java-parser'

class ParseImportsByCode extends BaseJavaCstVisitorWithDefaults {
  public data: Array<{
    type: 'static' | 'all'
    from: string
    flag: string
  }>
}

class ParseConsts extends BaseJavaCstVisitorWithDefaults {
  public data: Array<{
    name: string
    maybeValue:
      | {
          type: 'raw'
          value: any
        }
      | {
          type: 'ref'
          value: string[]
        }
  }>
}

export function parseImportsByCode(code: string) {
  const v = new ParseImportsByCode()
  v.visit(parse(code))
  return v.data
}

export function parseConstsByCode(code: string) {
  const v = new ParseConsts()
  v.visit(parse(code))
  return v.data
}

export function parseByCode(code: string) {
  return {
    imports: parseImportsByCode(code),
    consts: parseConstsByCode(code)
  }
}

// export async function parseConsts(path: string) {
//   const filenames = await globby(['**/*.java'], { cwd: nps.resolve(path), absolute: false, onlyFiles: true })
//
//   const obj = {}
//   await Promise.all(
//     filenames.map(async (name) => {
//       obj[name.replace(/\..+?$/, '')] = await parseByCode(fsExtra.readFile(nps.resolve(name), 'utf8'))
//     })
//   )
//   return obj
// }
