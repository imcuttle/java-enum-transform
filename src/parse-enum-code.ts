import {
  CstNode,
  BaseJavaCstVisitorWithDefaults,
  BaseJavaCstVisitor,
  parse,
  ClassDeclarationCtx,
  EnumDeclarationCtx,
  EnumConstantCtx,
  ExpressionCtx,
  IntegerLiteralCstNode,
  IntegerLiteralCtx,
  IToken
} from 'java-parser'

class ExpressionValueVisitor extends BaseJavaCstVisitorWithDefaults {
  public data: {
    type: 'fqnOrRefType' | 'primitive'
    raw: string
    value: any
  }
  setData(raw: any, type = 'primitive') {
    let value = raw
    if (type === 'primitive') {
      value = eval(raw)
    }
    this.data = {
      value,
      type,
      raw
    } as any
  }
  integerLiteral(ctx) {
    if (ctx.DecimalLiteral) {
      this.setData(ctx.DecimalLiteral[0].image)
    }
    if (ctx.HexLiteral) {
      this.setData(ctx.HexLiteral[0].image)
    }
    if (ctx.BinaryLiteral) {
      this.setData(ctx.BinaryLiteral[0].image)
    }
    if (ctx.OctalLiteral) {
      this.setData(ctx.OctalLiteral[0].image)
    }
    return super.integerLiteral(ctx)
  }
  floatingPointLiteral(ctx) {
    if (ctx.FloatLiteral) {
      this.setData(ctx.FloatLiteral[0].image)
    }
    if (ctx.HexFloatLiteral) {
      this.setData(ctx.HexFloatLiteral[0].image)
    }
    return super.floatingPointLiteral(ctx)
  }
  booleanLiteral(ctx) {
    if (ctx.True) {
      this.setData('true')
    }
    if (ctx.False) {
      this.setData('false')
    }
  }
  literal(ctx, param?: any): any {
    if (ctx.StringLiteral) {
      this.setData(ctx.StringLiteral[0].image)
    }
    if (ctx.Null) {
      this.setData('null')
    }
    if (ctx.CharLiteral) {
      this.setData(ctx.CharLiteral[0].image)
    }
    return super.literal(ctx, param)
  }

  fqnOrRefType(ctx) {
    const firstImg = ctx.fqnOrRefTypePartFirst[0].children.fqnOrRefTypePartCommon[0].children.Identifier[0].image
    const imgs = ctx.fqnOrRefTypePartRest.map(
      (rest) => rest.children.fqnOrRefTypePartCommon[0].children.Identifier[0].image
    )
    const value = [firstImg].concat(imgs)
    this.data = {
      type: 'fqnOrRefType',
      value,
      raw: JSON.stringify(value.join('.'))
    }
  }
}

function getExpressionValue(cst: any) {
  const v = new ExpressionValueVisitor()
  v.visit(cst)
  return v.data
}

class Visitor extends BaseJavaCstVisitorWithDefaults {
  public data: Array<{
    enumClass: string
    enums: Array<{ name: string; label?: ExpressionValueVisitor['data']; value: ExpressionValueVisitor['data'] }>
  }> = []

  public state = {
    visiting: false,
    enumClass: null,
    enums: []
  }

  startVisit(fn?) {
    this.state.visiting = true

    if (fn) {
      const res = fn()
      this.stopVisit()
      return res
    }
  }
  stopVisit() {
    if (this.state.visiting && this.state.enumClass) {
      this.data.push({
        enumClass: this.state.enumClass,
        enums: this.state.enums
      })
    }

    this.state.visiting = false
    this.state.enumClass = null
    this.state.enums = []
  }
  isVisiting() {
    return this.state.visiting
  }

  classDeclaration(ctx: ClassDeclarationCtx, param?: any): any {
    /**
     * 和服务端经过沟通，发现这种写法不太出现，暂时先不考虑支持嵌套内部类 enum
     * @example
     * public class RoleParent {
     *  class X {
     *   public enum Type {
     *     I_DEVELOPER(1, "开发者"),
     *     I_ADMIN(2,"管理"),
     *     I_USER(3,"用户");
     *   }
     *  }
     * }
     */
    if (this.isVisiting()) {
      return
    }

    if (ctx.classModifier?.[0]?.children?.Public) {
      if (ctx.enumDeclaration?.length) {
        return this.startVisit(() => super.classDeclaration(ctx, param))
      }
    }
  }

  enumDeclaration(ctx: EnumDeclarationCtx, param?: any): any {
    if (this.isVisiting()) {
      this.state.enumClass = ctx.typeIdentifier[0].children.Identifier[0].image
      return super.enumDeclaration(ctx, param)
    }
  }

  enumConstant(ctx: EnumConstantCtx, param?: any): any {
    if (this.isVisiting()) {
      let value
      let label
      if (
        // 两个参数
        // eg. `USER(1, "用户")`
        (ctx.argumentList[0].children?.expression?.length === 2 ||
          ctx.argumentList[0].children?.expression?.length === 1) &&
        (value = getExpressionValue(ctx.argumentList[0].children.expression[0]))
      ) {
        label =
          ctx.argumentList[0].children.expression[1] && getExpressionValue(ctx.argumentList[0].children.expression[1])
        this.state.enums.push({
          name: ctx.Identifier[0].image,
          label,
          value
        })
      }
    }
  }
}

export function parseEnumCode(codeOrAst: string | CstNode) {
  const ast = typeof codeOrAst === 'string' ? parse(codeOrAst) : codeOrAst
  const visitor = new Visitor()
  visitor.visit(ast)
  return visitor.data
}

export function formatToTs(data: Visitor['data']) {
  return data
    .map((x) => {
      const enumBodyString = x.enums
        .map((enumData) => {
          if (enumData.value.type === 'primitive') {
            return `  ${enumData.name} = ${enumData.value.raw},`
          }
          return ''
        })
        .filter(Boolean)
        .join('\n')
      if (!enumBodyString) {
        return ''
      }
      return `
export enum ${x.enumClass} {
${enumBodyString}
}
`.trim()
    })
    .filter(Boolean)
    .join('\n')
}
