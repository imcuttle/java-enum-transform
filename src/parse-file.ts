import * as nps from 'path'
import { CstNode, BaseJavaCstVisitorWithDefaults, parse } from 'java-parser'
import { sep, dirname } from 'path'

export interface Meta {
  type: 'enum' | 'class' | 'record' | null
  access: 'public' | 'protected' | 'private' | 'default'
  package: null | string[]
  name: string
}

interface ParseFileMetaOptions {
  classPath?: string
  filename?: string
}

class ParseFileMeta extends BaseJavaCstVisitorWithDefaults {
  constructor(protected options: ParseFileMetaOptions = {}) {
    super()
    const { classPath, filename } = this.options

    if (filename) {
      const relativeFilename = classPath ? nps.relative(classPath, filename) : filename
      this.data.package = dirname(relativeFilename).split(sep)
      this.data.name = nps.basename(relativeFilename, nps.extname(relativeFilename))
    }
  }

  public data: Meta = {
    type: null,
    access: 'default',
    package: null,
    name: ''
  }

  state = {
    inClassDeclaration: false,
    done: false
  }

  packageDeclaration(ctx) {
    if (ctx.Identifier?.length) {
      this.data.package = ctx.Identifier.map((x) => x.image)
    }
  }

  classDeclaration(ctx, param) {
    if (this.state.done || this.state.inClassDeclaration) {
      return
    }

    if (ctx.enumDeclaration) {
      this.data.type = 'enum'
    } else if (ctx.recordDeclaration) {
      this.data.type = 'record'
    } else if (ctx.normalClassDeclaration) {
      this.data.type = 'class'
    }
    const classModifierCtx = ctx.classModifier?.[0]?.children
    if (classModifierCtx) {
      if (classModifierCtx.Public) {
        this.data.access = 'public'
      } else if (classModifierCtx.Protected) {
        this.data.access = 'protected'
      } else if (classModifierCtx.Private) {
        this.data.access = 'private'
      }
      // classModifierCtx.Abstract
      // classModifierCtx.Final
    }

    this.state.inClassDeclaration = true
    return super.classDeclaration(ctx, param)
  }

  typeIdentifier(ctx, param) {
    if (!this.state.done && this.state.inClassDeclaration) {
      this.state.done = true
      this.data.name = ctx.Identifier[0]?.image || this.data.name
    }
  }
}

export function parseFileMeta(contentOrAst: string | CstNode, opts?: ParseFileMetaOptions) {
  const v = new ParseFileMeta(opts)
  v.visit(typeof contentOrAst === 'string' ? parse(contentOrAst) : contentOrAst)
  return v.data
}
