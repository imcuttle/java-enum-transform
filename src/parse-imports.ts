import { CstNode, BaseJavaCstVisitorWithDefaults, parse } from 'java-parser'

export interface Import {
  static: boolean
  package: string[]
  name?: string
  member: string[]
  wildcard: boolean
}

class ParseImport extends BaseJavaCstVisitorWithDefaults {
  public data: Import[] = []

  importDeclaration(ctx) {
    const importData: Import = {} as any
    importData.static = !!ctx.Static
    importData.wildcard = !!ctx.Star
    const images = ctx.packageOrTypeName[0].children.Identifier.map((x) => x.image)
    const nameIndex = images.findIndex((x) => /^[A-Z]/.test(x))
    if (nameIndex >= 0) {
      importData.package = images.slice(0, nameIndex)
      importData.name = images[nameIndex]
      importData.member = images.slice(nameIndex + 1)
    } else {
      importData.package = images
      importData.member = []
    }
    this.data.push(importData)
    // return super.importDeclaration(ctx)
  }
}

export function parseImport(contentOrAst: string | CstNode) {
  const v = new ParseImport()
  v.visit(typeof contentOrAst === 'string' ? parse(contentOrAst) : contentOrAst)
  return v.data
}
