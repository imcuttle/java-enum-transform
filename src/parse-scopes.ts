import * as nps from 'path'

export function parseDeclaresByCode(code: string) {}

export function parseDeclares(path: string) {
  path = nps.resolve(path)
}
