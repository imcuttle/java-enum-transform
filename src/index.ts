/**
 * Java enum transform
 * @author 余聪
 */
import * as globby from 'globby'
import * as fsExtra from 'fs-extra'
import { CstNode, parse } from 'java-parser'
import escape from 'escape-string-regexp'
import { delimiter, dirname } from 'path'
import { parseEnumCode } from './parse-enum-code'
import { Meta, parseFileMeta } from './parse-file'

export async function javaEnumTransformByDir(dir: string) {
  const filenames = await globby('**/*.java', {
    absolute: false,
    onlyFiles: true,
    cwd: dir
  })

  const fileMap = new Map<string, { meta: Meta; code: string; package: string; ast: CstNode }>()
  filenames.map(async (filename) => {
    const code = await fsExtra.readFile(filename, 'utf-8')
    const ast = parse(code)
    const meta = parseFileMeta(ast)

    const pkg = meta.package?.length
      ? meta.package.join('.')
      : dirname(filename).replace(new RegExp(escape(delimiter), 'g'), '.')

    if (pkg) {
      fileMap.set(pkg, {
        meta,
        code,
        ast,
        package: pkg
      })
    }
  })

  const result = new Map<string, ReturnType<typeof parseEnumCode>>()
  await Promise.all(
    Array.from(fileMap.values())
      .filter((file) => file.meta.type === 'enum')
      .map(async (file) => {
        const list = parseEnumCode(file.ast)
        result.set(file.package, list)
      })
  )
  return result
}
