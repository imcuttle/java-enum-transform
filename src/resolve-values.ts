import {
  CstNode,
  BaseJavaCstVisitorWithDefaults,
  parse,
  ClassDeclarationCtx,
  EnumDeclarationCtx,
  EnumConstantCtx,
  ExpressionCtx,
  IntegerLiteralCstNode,
  IntegerLiteralCtx,
  IToken
} from 'java-parser'
import { parseImport } from './parse-imports'
import { getExpressionValue } from './parse-enum-code'

/**
 * 解析
 * class class X {
 *   public static int a = 123;
 *   private boolean b = true;
 * }
 * // => { a: { type: 'primitive', value: 123, raw: '123' }, b: { type: 'primitive', value: true, raw: 'true' } }
 */
class ParseBodyMember extends BaseJavaCstVisitorWithDefaults {
  public data: Map<string, { value: ReturnType<typeof getExpressionValue>; static: boolean }> = new Map()

  protected state = {
    inClassMember: false
  }

  fieldDeclaration(ctx) {
    this.state.inClassMember = true
    super.classMemberDeclaration(ctx)
    this.state.inClassMember = false

    const isStatic = !!ctx.fieldModifier.find((x) => x.children.Static)
    const tmpCtx = ctx.variableDeclaratorList[0].children.variableDeclarator[0].children

    if (tmpCtx.variableInitializer && tmpCtx.variableDeclaratorId && tmpCtx.Equals) {
      const value = getExpressionValue(tmpCtx.variableInitializer[0].children.expression[0])
      if (value) {
        this.data.set(tmpCtx.variableDeclaratorId[0].children.Identifier[0].image, {
          value,
          static: isStatic
        })
      }
    }
  }
}

export function parseBodyMember(ast: any) {
  const v = new ParseBodyMember()
  v.visit(ast)
  return v.data
}

/**
 * 解决嵌套，动态计算的值
 * @param ast
 * @param getAstMapByPackage
 * @param getAstById
 * @param resolveExpressionValue
 * @param map
 */
export const resolveImportsValue = (
  ast: any,
  {
    getAstMapByPackage,
    getAstById,
    map = new Map(),
    resolveExpressionValue
  }: {
    map?: Map<any, ReturnType<typeof parseBodyMember>>
    getAstById: (id: string) => any
    getAstMapByPackage: (packageId: string) => Record<string, any>
    resolveExpressionValue?: (
      item: ReturnType<typeof getExpressionValue>,
      id: string,
      fieldName: string
    ) => ReturnType<typeof getExpressionValue>
  }
): null | Map<string, ParseBodyMember['data'] | { value: ReturnType<typeof getExpressionValue>; static: boolean }> => {
  if (!ast) {
    return null
  }
  const imports = parseImport(ast)
  const scopedMap = new Map<
    string,
    ParseBodyMember['data'] | { value: ReturnType<typeof getExpressionValue>; static: boolean }
  >()

  const getMemberMap = (id) => {
    const newAst = getAstById(id)
    if (!newAst) {
      return
    }

    if (map.get(id)) {
      return map.get(id)
    }
    const memberMap = parseBodyMember(newAst)
    if (resolveExpressionValue) {
      memberMap.forEach((item, key) => {
        item.value = resolveExpressionValue(item.value, id, key)
      })
    }
    map.set(id, memberMap)
    return memberMap as ParseBodyMember['data']
  }
  const getMemberMapByPackage = (id) => {
    const result = new Map()
    const newAsts = getAstMapByPackage(id)
    if (!newAsts) {
      return result
    }
    Object.keys(newAsts).forEach((name) => {
      const classid = [id].concat(name).join('.')
      result.set(name, getMemberMap(classid))
    })
    return result
  }

  imports.forEach((x) => {
    const id = x.package ? x.package.concat(x.name).filter(Boolean).join('.') : x.name
    if (!id) {
      return
    }

    if (x.wildcard) {
      if (x.name) {
        // import java.io.Util.*;
        const memberMap = getMemberMap(id)
        if (memberMap) {
          for (const [name, item] of memberMap.entries()) {
            if (item.static) {
              scopedMap.set(name, item)
            }
          }
        }
      } else {
        // import java.io.*;
        const packageMap = getMemberMapByPackage(id)
        if (packageMap) {
          for (const [name, item] of packageMap.entries()) {
            scopedMap.set(name, item)
          }
        }
      }
      return
    }

    if (x.name) {
      // import java.io.Util.a;
      if (x.member?.length) {
        if (x.member.length === 1) {
          const memberMap = getMemberMap(id)
          if (memberMap && memberMap.get(x.member[0])) {
            scopedMap.set(x.member[0], memberMap.get(x.member[0]))
          }
        }
      } else {
        // import java.io.Util;
        const memberMap = getMemberMap(id)
        if (memberMap) {
          scopedMap.set(x.name, memberMap)
        }
      }
    }
  })
  return scopedMap
}

export function resolveExpressionValue(
  item: ReturnType<typeof getExpressionValue>,
  {
    _visitMap = new Map(),
    id,
    fieldName,
    ...opts
  }: Omit<Parameters<typeof resolveImportsValue>[1], 'resolveExpressionValue'> & {
    _visitMap?: Map<any, any>
    id: string
    fieldName: string
  }
) {
  const options = {
    _visitMap,
    id,
    fieldName,
    ...opts
  }

  if (item.type === 'fqnOrRefType') {
    const fieldKey = `${id}#${fieldName}`
    const prevData = _visitMap.get(fieldKey)
    if (prevData?.status === 'post') {
      return prevData.data
    }

    if (prevData?.status === 'pre') {
      throw new Error(`存在环形依赖解析 ${fieldKey}`)
    }

    _visitMap.set(fieldKey, { status: 'pre' })
    const importsValue = resolveImportsValue(opts.getAstById(id), {
      ...opts,
      resolveExpressionValue: (item, id, key) => {
        return resolveExpressionValue(item, { ...options, fieldName: key, id })
      }
    })
    if (!importsValue) {
      _visitMap.set(fieldKey, { status: 'post', data: item })
      return item
    }

    if (item.value?.length) {
      let map: any = importsValue
      for (let i = 0; i < item.value.length; i++) {
        map = map.get(item.value[i])
        if (!(map instanceof Map)) {
          if (i === item.value.length - 1) {
            _visitMap.set(fieldKey, { status: 'post', data: map?.value ?? item })
            return map?.value ?? item
          }
          break
        }
      }
    }
  }
  return item
}
