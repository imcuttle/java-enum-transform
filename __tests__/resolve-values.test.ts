/**
 * @file main
 * @author imcuttle
 * @date 2018/4/4
 */
import { parse } from 'java-parser'
import { parseBodyMember, resolveExpressionValue, resolveImportsValue } from '../src/resolve-values'

const astMap = {
  'com.example.Main': parse(`import com.example.utils.*;
import com.example.Foo;
import com.example.Foo.*;`),
  'com.example.Foo': parse(`import static com.example.utils.B.b;

public class Foo {
  public static int count = 1;
  public static int count_invalid;
  public static boolean isOk = true;
  public static boolean isOk_invalid;
  public String name = "imcuttle";
  public String concatString = "a" + "b";
  public String computed_1 = b;
  public String computed_2 = b + "tail";
}
      `),
  'com.example.utils.A': parse(`
public class A {
  public static char a = 'a';
}
      `),
  'com.example.utils.B': parse(`
public class A {
  public static String b = "b";
}
      `)
}

const getAstById = jest.fn((id) => {
  return astMap[id]
})
const getAstMapByPackage = jest.fn((scope) => {
  const asts = {}
  Object.keys(astMap).forEach((name) => {
    if (name.startsWith(scope + '.') && !name.slice((scope + '.').length).includes('.')) {
      asts[name.slice((scope + '.').length)] = astMap[name]
    }
  })
  return asts
})

describe('resolve-values', function () {
  beforeEach(() => {
    getAstById.mockClear()
    getAstMapByPackage.mockClear()
  })

  it('parseBodyMember', function () {
    expect(
      parseBodyMember(
        parse(`
    import static com.example.utils.B.b;

public class Foo {
  public static int count = 1;
  public static int count_invalid;
  public static boolean isOk = true;
  public static boolean isOk_invalid;
  public String name = "imcuttle";
  // 暂不支持
  public String concatString = "a" + "b";
  public String computed_1 = b;
  // 暂不支持
  public String computed_2 = b + "tail";
}
`)
      )
    ).toMatchInlineSnapshot(`
      Map {
        "count" => Object {
          "static": true,
          "value": Object {
            "raw": "1",
            "type": "primitive",
            "value": 1,
          },
        },
        "isOk" => Object {
          "static": true,
          "value": Object {
            "raw": "true",
            "type": "primitive",
            "value": true,
          },
        },
        "name" => Object {
          "static": false,
          "value": Object {
            "raw": "\\"imcuttle\\"",
            "type": "primitive",
            "value": "imcuttle",
          },
        },
        "computed_1" => Object {
          "static": false,
          "value": Object {
            "raw": "b",
            "type": "fqnOrRefType",
            "value": Array [
              "b",
            ],
          },
        },
      }
    `)
  })

  it('resolveImportsValue', async function () {
    const result = resolveImportsValue(astMap['com.example.Main'], {
      getAstById,
      getAstMapByPackage
    })
    expect(result).toMatchSnapshot()
    // same ref
    expect(result.get('count')).toBe((result.get('Foo') as any).get('count'))

    expect(getAstById.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "com.example.utils.A",
        ],
        Array [
          "com.example.utils.B",
        ],
        Array [
          "com.example.Foo",
        ],
        Array [
          "com.example.Foo",
        ],
      ]
    `)
    expect(getAstMapByPackage.mock.calls).toMatchInlineSnapshot(`
      Array [
        Array [
          "com.example.utils",
        ],
      ]
    `)
  })
  it('resolveImportsValue 1', async function () {
    const result = resolveImportsValue(
      parse(`
    import static com.example.utils.B.b;
    `),
      {
        getAstById,
        getAstMapByPackage
      }
    )
    expect(result).toMatchInlineSnapshot(`
Map {
  "b" => Object {
    "static": true,
    "value": Object {
      "raw": "\\"b\\"",
      "type": "primitive",
      "value": "b",
    },
  },
}
`)
  })

  it('resolveExpressionValue#primitive', function () {
    const opts = {
      getAstById,
      getAstMapByPackage,
      id: 'com.example.Foo',
      fieldName: 'bar'
    }

    const item = {
      value: true,
      raw: 'true',
      type: 'primitive'
    }
    expect(resolveExpressionValue(item as any, opts)).toBe(item)
  })

  it('resolveExpressionValue#fqnOrRefType', function () {
    const opts = {
      getAstById,
      getAstMapByPackage,
      id: 'com.example.Main',
      fieldName: 'foo'
    }

    const item: Parameters<typeof resolveExpressionValue>[0] = {
      value: ['Foo', 'count'],
      raw: 'Foo.count',
      type: 'fqnOrRefType'
    }

    expect(resolveExpressionValue(item, opts)).toEqual({
      raw: '1',
      type: 'primitive',
      value: 1
    })
  })

  it('resolveExpressionValue#fqnOrRefType computed', function () {
    const opts = {
      getAstById,
      getAstMapByPackage,
      id: 'com.example.Main',
      fieldName: 'foo'
    }

    const item: Parameters<typeof resolveExpressionValue>[0] = {
      value: ['Foo', 'computed_1'],
      raw: 'Foo.computed_1',
      type: 'fqnOrRefType'
    }

    expect(resolveExpressionValue(item, opts)).toEqual({
      raw: '"b"',
      type: 'primitive',
      value: 'b'
    })
  })
})
