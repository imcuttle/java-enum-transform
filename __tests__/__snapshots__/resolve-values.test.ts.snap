// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolve-values resolveImportsValue 1`] = `
Map {
  "A" => Map {
    "a" => Object {
      "static": true,
      "value": Object {
        "raw": "'a'",
        "type": "primitive",
        "value": "a",
      },
    },
  },
  "B" => Map {
    "b" => Object {
      "static": true,
      "value": Object {
        "raw": "\\"b\\"",
        "type": "primitive",
        "value": "b",
      },
    },
  },
  "Foo" => Map {
    "count" => Object {
      "static": true,
      "value": Object {
        "raw": "1",
        "type": "primitive",
        "value": 1,
      },
    },
    "isOk" => Object {
      "static": true,
      "value": Object {
        "raw": "true",
        "type": "primitive",
        "value": true,
      },
    },
    "name" => Object {
      "static": false,
      "value": Object {
        "raw": "\\"imcuttle\\"",
        "type": "primitive",
        "value": "imcuttle",
      },
    },
    "computed_1" => Object {
      "static": false,
      "value": Object {
        "raw": "b",
        "type": "fqnOrRefType",
        "value": Array [
          "b",
        ],
      },
    },
  },
  "count" => Object {
    "static": true,
    "value": Object {
      "raw": "1",
      "type": "primitive",
      "value": 1,
    },
  },
  "isOk" => Object {
    "static": true,
    "value": Object {
      "raw": "true",
      "type": "primitive",
      "value": true,
    },
  },
}
`;
